File: internal/session/session.go
// Package session holds the state machine (kb:anchor/state) and the in-memory
// session registry that feeds it. It operates purely on claudecode.StateInput values
// and its own neutral fields — no Claude Code payload key or event name appears here
// (CLAUDE.md hard rule; plan m1-sessions "The state machine — implementation shape").
package session

import (
	"time"
	"unicode/utf8"

	"github.com/Zalaras/muster/internal/claudecode"
)

// State is one of the six displayed states (kb:anchor/state.displayed).
type State string

const (
	StateStarted    State = "started"
	StatePlanning   State = "planning"
	StateWorking    State = "working"
	StateNeedsInput State = "needs_input"
	StateFailed     State = "failed"
	StateIdle       State = "idle"
)

// PermissionMode is the latched last-known permission mode (kb:anchor/state.tracked).
// The four values are Claude Code's own `--permission-mode` vocabulary
// (kb:fact/permission-mode-no-flag-follows-configured-default), owned by
// internal/claudecode (CLAUDE.md hard rule; maintainability-cleanup review Major
// 6/c-Minor 11) — this package's constants derive from that owner rather than
// re-declaring the literals, since session already imports claudecode (for StateInput),
// so this direction never cycles.
type PermissionMode string

const (
	PermissionDefault     PermissionMode = PermissionMode(claudecode.PermissionDefault)
	PermissionPlan        PermissionMode = PermissionMode(claudecode.PermissionPlan)
	PermissionAcceptEdits PermissionMode = PermissionMode(claudecode.PermissionAcceptEdits)
	// PermissionAuto is Claude Code's distinct "auto" mode, measured 2026-09-03 against
	// 2.1.259 (docs/history/spikes/canary-fields.md § Hook payloads, "Permission-mode probe"): reports
	// permission_mode "auto" on hooks; model-gated (falls back to "default" on haiku).
	PermissionAuto PermissionMode = PermissionMode(claudecode.PermissionAuto)
)

// PermissionModes lists every mode as this package's own typed enum — internal/server
// validates incoming requests against it rather than re-spelling the four literals
// itself (maintainability-cleanup review, Major 6).
var PermissionModes = []PermissionMode{PermissionDefault, PermissionPlan, PermissionAcceptEdits, PermissionAuto}

// ValidPermissionMode reports whether s is one of PermissionModes. Delegates to
// claudecode.ValidPermissionMode, the one owner of the underlying set, rather than
// walking PermissionModes itself — one validation, not two.
func ValidPermissionMode(s string) bool {
	return claudecode.ValidPermissionMode(s)
}

// Attention is non-nil iff State == StateNeedsInput.
type Attention struct {
	Reason string // "permission" | "idle"
	Since  time.Time
}

// Failure is non-nil iff State == StateFailed.
type Failure struct {
	Error   string // raw token, displayed verbatim, never switched on
	Message string
}

// Model is the session's model readout (kb:anchor/ws.session value semantics); both
// fields start at the launch value verbatim, and a routed status-line post refreshes both
// whenever its model object is present.
type Model struct {
	ID          string
	DisplayName string
}

// Context is the session's live context-window gauge (kb:anchor/ws.session value
// semantics): UsedPct/TotalInputTokens/WindowSize are always all present together — a nil
// *Context means unknown, never a zero value. Populated only by a routed status-line post
// whose payload carries a non-null used-percentage; reset to nil by `/clear`.
type Context struct {
	UsedPct          float64
	TotalInputTokens int64
	WindowSize       int64
}

// Session is one row of the kb:anchor/state state machine, held in memory and persisted on every
// mutation. Guard fields (currentPromptID/closedPromptIDs) are in-memory only — a
// daemon restart resets them.
type Session struct {
	ID                   int64
	TmuxTarget           string
	TmuxPane             string
	ClaudeSessionID      string // "" until bound
	RepoID               int64
	Directory            string
	Branch               *string // nil iff Directory isn't a git checkout
	IsWorktree           bool
	Title                *string // Claude's last-known name (status-line session_name, or launch --name until then)
	State                State
	StateSince           time.Time
	PermissionMode       PermissionMode
	PermissionModeSource string // "seed" | "hook"
	Model                *Model
	Context              *Context
	Compactions          int
	Attention            *Attention
	Failure              *Failure
	LastActivity         *string
	Alive                bool
	EndedAt              *time.Time
	FirstLaunchHere      bool
	CreatedAt            time.Time

	// LastSnapshot/LastSnapshotAt (m4-reconcile REQ-4): the last capture-pane text and
	// when it was captured. "" / zero means "never captured yet" — display source only,
	// never read by the state machine, never logged (may hold prompt text).
	LastSnapshot   string
	LastSnapshotAt time.Time

	// Pinned/RailPos (plan order-sidebar): user-owned rail order. Display-only —
	// never read by the state machine (D17); mutated only by railorder.go's pure
	// applyPin/applyOrder, via Manager.SetPinned/SetOrder.
	Pinned  bool
	RailPos int64

	// TitleOverride (plan ui-text-and-focus REQ-9/REQ-11): the user's rename via PUT
	// .../title, nil = none. Display-only, wins over Title in DisplayTitle() — never
	// read by the state machine or the status path (INV-2); mutated only by
	// Manager.SetTitle.
	TitleOverride *string

	// TranscriptPath/PlanPath/PlanExists (plan markdown-viewing REQ-16/REQ-17): the
	// latest transcript path a routed hook named, and the plan derived from it via
	// claudecode.LocatePlanFile. Display-only, never read by machine.go. PlanPath ""
	// (the wire plan:null) means no transcript has ever named a plan; once a plan has
	// been named, a planless scan keeps it rather than clearing it back to ""
	// (kb:adr/reader-plan-sticky-once-named). Mutated only by
	// Manager.SetTranscript/SetPlan/ApplyPlanScan.
	TranscriptPath string
	PlanPath       string
	PlanExists     bool

	// Unread (plan rail-card-improvements REQ-7): true iff the turn closed with no
	// terminal client attached to this session, since cleared. INV: Unread ⇒ State ==
	// idle, enforced by setState — the one place every applyInput arm routes through.
	// Set by Manager.Apply after a turn_closed input per the watcher's answer; cleared by
	// Manager.MarkSeen (attach on either surface). Display-only, independent of Alive,
	// survives a restart.
	Unread bool

	// LastPrompt (plan rail-card-improvements REQ-12): the user's most recent prompt,
	// truncated to 200 chars, nil until a first prompt or after /clear. Set by
	// KindTurnActivity when the adapter supplied one; display-only, never read by the
	// state machine.
	LastPrompt *string

	currentPromptID string
	closedPromptIDs []string // bounded ring, most recent last, capped at maxClosedPrompts
}

const maxClosedPrompts = 8

// Clone returns a value copy safe to hand outside the manager's lock (Attention,
// Failure and Model are pointers to otherwise-immutable snapshots — callers must treat
// them as read-only).
func (s *Session) Clone() *Session {
	c := *s
	return &c
}

// DisplayTitle returns the wire "title" (plan ui-text-and-focus REQ-11): TitleOverride
// when non-nil, else Claude's last-known name (Title), else nil. The daemon owns this
// precedence; no client computes it (kb:anchor/ws.session).
func (s *Session) DisplayTitle() *string {
	if s.TitleOverride != nil {
		return s.TitleOverride
	}
	return s.Title
}

func (s *Session) promptClosed(id string) bool {
	for _, closed := range s.closedPromptIDs {
		if closed == id {
			return true
		}
	}
	return false
}

func (s *Session) closePrompt(id string) {
	if s.promptClosed(id) {
		return
	}
	s.closedPromptIDs = append(s.closedPromptIDs, id)
	if len(s.closedPromptIDs) > maxClosedPrompts {
		s.closedPromptIDs = s.closedPromptIDs[len(s.closedPromptIDs)-maxClosedPrompts:]
	}
	if s.currentPromptID == id {
		s.currentPromptID = ""
	}
}

// setState transitions to next, updating stateSince only when the state actually
// changes (Edge Case 3: reapplying an event to the same state is a no-op transition).
// Every transition to a non-idle state clears Unread (kb:anchor/state.tracked's Unread ⇒
// idle invariant) — enforced here, the one place every applyInput arm routes through, so
// no arm can strand it. Manager.Apply is the only place that ever sets Unread true, after
// applyInput returns.
func (s *Session) setState(next State, now time.Time) {
	if next != StateIdle {
		s.Unread = false
	}
	if s.State == next {
		return
	}
	s.State = next
	s.StateSince = now
}

// activeState is "planning" if the permission-mode latch reads "plan", else "working"
// (kb:anchor/state.transitions's ACTIVE definition).
func (s *Session) activeState() State {
	if s.PermissionMode == PermissionPlan {
		return StatePlanning
	}
	return StateWorking
}

// stringPtrEqual reports whether two nullable strings hold the same value — nil equals
// only nil (plan ui-text-and-focus: SetTitle/applyStatusUpdate's "did the wire title
// change" checks).
func stringPtrEqual(a, b *string) bool {
	if a == nil || b == nil {
		return a == nil && b == nil
	}
	return *a == *b
}

// truncate cuts s to at most n bytes without splitting a multi-byte UTF-8 rune:
// LastPrompt/LastActivity are hook-supplied text, and a byte-count cut that lands mid-rune
// would persist and broadcast invalid UTF-8.
func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	for n > 0 && !utf8.RuneStart(s[n]) {
		n--
	}
	return s[:n]
}

File: web/src/protocol/session.ts
// Wire type and parser for one `Session` (docs/protocol.md, kb:anchor/ws.session) — one of
// the protocol/ concept modules, each owning one wire concept's type and parser;
// `decode.ts` holds the primitives every one of them shares.

import { asNumber, isRecord, parseNullable } from "./decode";

export const SESSION_STATES = [
  "started",
  "planning",
  "working",
  "needs_input",
  "failed",
  "idle",
] as const;
export type SessionState = (typeof SESSION_STATES)[number];

export const ATTENTION_REASONS = ["permission", "idle"] as const;
export type AttentionReason = (typeof ATTENTION_REASONS)[number];

export interface SessionAttention {
  reason: AttentionReason;
  since: string;
}

export interface SessionFailure {
  error: string;
  message: string;
}

export interface SessionRepo {
  name: string;
  branch: string | null;
  isWorktree: boolean;
}

export interface SessionModelInfo {
  id: string;
  displayName: string;
}

export const PERMISSION_MODE_SOURCES = ["seed", "hook"] as const;
export type PermissionModeSource = (typeof PERMISSION_MODE_SOURCES)[number];

export interface PermissionModeInfo {
  value: string;
  source: PermissionModeSource;
}

// kb:anchor/sessions.create: the accepted wire values for `permissionMode` on `POST
// /api/sessions`, in dialog/cycle order. `default` is Claude Code's manual mode (the UI
// labels it "manual"); one source for both the request union and the launch dialog's radio
// guard, so they can't drift apart.
export const PERMISSION_MODES = ["default", "acceptEdits", "plan", "auto"] as const;
export type PermissionMode = (typeof PERMISSION_MODES)[number];

export function isPermissionMode(value: unknown): value is PermissionMode {
  return (PERMISSION_MODES as readonly unknown[]).includes(value);
}

export interface SessionContext {
  usedPct: number | null;
  totalInputTokens: number | null;
  windowSize: number | null;
  compactions: number;
}

// kb:anchor/ws.session: the plan the transcript scan derived.
// `null` until a transcript of this session has named a plan (never entered plan mode);
// once non-null it is never null again for the row's lifetime — a scan that finds no plan
// (a `/clear`'s fresh transcript, a deleted transcript) keeps the path and re-checks
// `exists`, and only a scan that names a plan replaces it (kb:adr/reader-plan-sticky-once-named).
// `exists: false` means the path is known but no file is there: plan mode entered with
// nothing written yet, or the file was since deleted.
export interface SessionPlan {
  path: string;
  exists: boolean;
}

// Full shape per kb:anchor/ws.session; parseSession below validates every field rather
// than trusting the daemon.
export interface Session {
  id: number;
  title: string | null;
  state: SessionState;
  stateSince: string;
  alive: boolean;
  endedAt: string | null;
  attention: SessionAttention | null;
  failure: SessionFailure | null;
  directory: string;
  repo: SessionRepo | null;
  model: SessionModelInfo | null;
  permissionMode: PermissionModeInfo;
  context: SessionContext;
  lastActivity: string | null;
  claudeSessionId: string | null;
  tmuxTarget: string;
  // kb:anchor/ws.session: true iff the launch created this directory's repo row —
  // drives the trust-prompt vs. no-signal honesty note client-side.
  firstLaunchHere: boolean;
  createdAt: string;
  // kb:anchor/ws.session: the rail's user-owned order. Required on every
  // wire Session (never null, both fields ship together) — unlike `model`/`usageModel`'s
  // "absent key defaults" pattern, no daemon version predates this field (the
  // daemon and client shipped it together), so parseSession below rejects a
  // session missing either field rather than defaulting it.
  pinned: boolean;
  railPos: number;
  // kb:anchor/ws.session: the user's rename via `PUT
  // /api/sessions/{id}/title`; `null` means no override. `title` above is already the
  // daemon's precedence-resolved DISPLAY title (titleOverride when non-null, else
  // Claude's last-known name) — a client renders `title` and reads this field only to
  // decide whether "clear" means anything (sessions/rename.ts's `titleCommand`).
  // Required on every wire Session, same "daemon and client shipped it together" reasoning as
  // pinned/railPos above.
  titleOverride: string | null;
  // kb:anchor/ws.session: additive, no protocol bump
  // (kb:adr/connection-protocol-bumps-only-on-shape-change) — but required on every wire
  // Session all the same, same "daemon and client shipped it together" reasoning as
  // pinned/railPos/titleOverride above.
  plan: SessionPlan | null;
  // kb:anchor/ws.session: true iff the turn closed
  // while no terminal client was attached to this session on either surface and nobody
  // has attached since. INV: unread ⇒ state == "idle". Required on every wire Session,
  // same "daemon and client shipped it together" reasoning as pinned/railPos/titleOverride/plan
  // above.
  unread: boolean;
  // kb:anchor/ws.session: the user's most recent
  // prompt, truncated to 200 chars; `null` until a first prompt and again after `/clear`.
  // Display-only — never logged (kb:adr/rail-activity-line-turn-aware-default-with-pref).
  // Same "ship together" reasoning as `unread` above.
  lastPrompt: string | null;
}

function isSessionState(value: unknown): value is SessionState {
  return (SESSION_STATES as readonly unknown[]).includes(value);
}

function isAttentionReason(value: unknown): value is AttentionReason {
  return (ATTENTION_REASONS as readonly unknown[]).includes(value);
}

function parseAttention(value: unknown): SessionAttention | null {
  if (!isRecord(value)) return null;
  const reason = value["reason"];
  const since = value["since"];
  if (!isAttentionReason(reason)) return null;
  if (typeof since !== "string") return null;
  return { reason, since };
}

function parseFailure(value: unknown): SessionFailure | null {
  if (!isRecord(value)) return null;
  const error = value["error"];
  const message = value["message"];
  if (typeof error !== "string" || typeof message !== "string") return null;
  return { error, message };
}

function parseRepoInfo(value: unknown): SessionRepo | null {
  if (!isRecord(value)) return null;
  const name = value["name"];
  const branch = value["branch"];
  const isWorktree = value["isWorktree"];
  if (typeof name !== "string") return null;
  if (branch !== null && typeof branch !== "string") return null;
  if (typeof isWorktree !== "boolean") return null;
  return { name, branch, isWorktree };
}

// Exported: usage.ts's `parseUsage` decodes the same shape for `Usage.model`
// (kb:anchor/ws.usage) rather than keeping its own copy.
export function parseModelInfo(value: unknown): SessionModelInfo | null {
  if (!isRecord(value)) return null;
  const id = value["id"];
  const displayName = value["displayName"];
  if (typeof id !== "string" || typeof displayName !== "string") return null;
  return { id, displayName };
}

function isPermissionModeSource(value: unknown): value is PermissionModeSource {
  return (PERMISSION_MODE_SOURCES as readonly unknown[]).includes(value);
}

function parsePermissionModeInfo(value: unknown): PermissionModeInfo | null {
  if (!isRecord(value)) return null;
  const modeValue = value["value"];
  const source = value["source"];
  if (typeof modeValue !== "string") return null;
  if (!isPermissionModeSource(source)) return null;
  return { value: modeValue, source };
}

function parseContext(value: unknown): SessionContext | null {
  if (!isRecord(value)) return null;
  const usedPct = parseNullable(value["usedPct"], asNumber);
  const totalInputTokens = parseNullable(value["totalInputTokens"], asNumber);
  const windowSize = parseNullable(value["windowSize"], asNumber);
  const compactions = value["compactions"];
  if (usedPct === undefined || totalInputTokens === undefined || windowSize === undefined)
    return null;
  if (typeof compactions !== "number") return null;
  return { usedPct, totalInputTokens, windowSize, compactions };
}

function parseSessionPlan(value: unknown): SessionPlan | null {
  if (!isRecord(value)) return null;
  const path = value["path"];
  const exists = value["exists"];
  if (typeof path !== "string") return null;
  if (typeof exists !== "boolean") return null;
  return { path, exists };
}

/** Validates one Session object per kb:anchor/ws.session. Every field is read-checked;
 * an unrecognized field name or type anywhere in the object rejects the whole session
 * (the caller drops the snapshot/upsert rather than render a half-formed card).
 *
 * Scores 41 on cognitive complexity (adding `unread`/
 * `lastPrompt`'s two guards), but every point is a flat `return null` guard at zero
 * nesting — the score tracks the wire object's field count, not any tangle. Splitting
 * it would scatter the "any bad field rejects the whole session" invariant across
 * several functions, where no single reader or test sees it whole. */
// biome-ignore lint/complexity/noExcessiveCognitiveComplexity: flat per-field guards; a split strands the all-or-nothing invariant
export function parseSession(value: unknown): Session | null {
  if (!isRecord(value)) return null;

  const id = value["id"];
  const title = value["title"];
  const state = value["state"];
  const stateSince = value["stateSince"];
  const alive = value["alive"];
  const endedAt = value["endedAt"];
  const rawAttention = value["attention"];
  const rawFailure = value["failure"];
  const directory = value["directory"];
  const rawRepo = value["repo"];
  const rawModel = value["model"];
  const lastActivity = value["lastActivity"];
  const claudeSessionId = value["claudeSessionId"];
  const tmuxTarget = value["tmuxTarget"];
  const firstLaunchHere = value["firstLaunchHere"];
  const createdAt = value["createdAt"];
  const pinned = value["pinned"];
  const railPos = value["railPos"];
  const titleOverride = value["titleOverride"];
  const rawPlan = value["plan"];
  const unread = value["unread"];
  const lastPrompt = value["lastPrompt"];

  if (typeof id !== "number") return null;
  if (title !== null && typeof title !== "string") return null;
  if (!isSessionState(state)) return null;
  if (typeof stateSince !== "string") return null;
  if (typeof alive !== "boolean") return null;
  if (endedAt !== null && typeof endedAt !== "string") return null;

  const attention = parseNullable(rawAttention, parseAttention);
  if (attention === undefined) return null;

  const failure = parseNullable(rawFailure, parseFailure);
  if (failure === undefined) return null;

  if (typeof directory !== "string") return null;

  const repo = parseNullable(rawRepo, parseRepoInfo);
  if (repo === undefined) return null;

  const model = parseNullable(rawModel, parseModelInfo);
  if (model === undefined) return null;

  const permissionMode = parsePermissionModeInfo(value["permissionMode"]);
  if (!permissionMode) return null;

  const context = parseContext(value["context"]);
  if (!context) return null;

  if (lastActivity !== null && typeof lastActivity !== "string") return null;
  if (claudeSessionId !== null && typeof claudeSessionId !== "string") return null;
  if (typeof tmuxTarget !== "string") return null;
  if (typeof firstLaunchHere !== "boolean") return null;
  if (typeof createdAt !== "string") return null;
  if (typeof pinned !== "boolean") return null;
  if (typeof railPos !== "number") return null;
  if (titleOverride !== null && typeof titleOverride !== "string") return null;

  const plan = parseNullable(rawPlan, parseSessionPlan);
  if (plan === undefined) return null;

  if (typeof unread !== "boolean") return null;
  if (lastPrompt !== null && typeof lastPrompt !== "string") return null;

  return {
    id,
    title,
    state,
    stateSince,
    alive,
    endedAt,
    attention,
    failure,
    directory,
    repo,
    model,
    permissionMode,
    context,
    lastActivity,
    claudeSessionId,
    tmuxTarget,
    firstLaunchHere,
    createdAt,
    pinned,
    railPos,
    titleOverride,
    plan,
    unread,
    lastPrompt,
  };
}
